FIx for #215

mbkumar opened the pull request:
This PR addresses multiple issues.

1) Fixes #215 by deprecating fix_all and adding fix_local and fix_full functions. Currently fix_all points to fix_local, but this behavior may be changed in the future by pointing it to fix_full.
2) Introduces a preliminary scaffolding for serializing Optimizable objects into JSON/YAML objects. Tested for individual Optimizable objects without any dependencies and for Optimizable objects with dependencies.
3) Source distribution now contains the external packages also. 
4) Modules with `graph_` prefix are deprecated and are supplanted with corresponding modules without the `graph_` prefix. They will be removed completely in the future.
Due to 4, this PR needs to be merged at the earliest to minimize merge conflicts when someone updates code in those modules.

mbkumar commented:
Git is not playing nice when a file is moved and then a new file with the same name is created. 
Changes for 1. and 2. are lost in optimizable.py, which shows up as a new file despite my best efforts.
For 2, look at `as_dict` and `from_dict` methods in Optimizable and a few subclasses such as `objectives.functions.Adder`.

landreman reviewed (commented):
- fix_all() and unfix_all() are still used in many unit tests and in some of the examples.
- The `docs/source/optimizable.rst` docs page still talks about `fix_all()` and `unfix_all()` rather than the new local/full functions.
- I see a few remaining `graph_` prefixes in `docs/source/mpi.rst`, `src/_core/optimizable.py`, and `examples/README.md`
- In tests/core there is both test_optimizable and test_graph_optimizable.
- In tests/objectives there is both test_least_squares and test_graph_least_squares
- In tests/solve there is both test_mpi and test_graph_mpi, and both test_least_squares and test_graph_least_squares
- There are several orthogonal changes going on in this PR. In the future it would be nice to have these kinds of changes as separate PRs. Then we could e.g. resolve #215 without having to simultaneously merge the incomplete serialization bits.

mbkumar commented on src/simsopt/objectives/functions.py:
```diff
@@ -0,0 +1,376 @@
+        d["@class"] = self.__class__.__name__
+        d["b"] = self._sqrtb * self._sqrtb
+        d["x"] = self.get("x")
+        d["y"] = self.get("y")
```
Thanks! Fixed it.

mbkumar commented:
Thanks, Matt for quickly going through the changes.

> * fix_all() and unfix_all() are still used in many unit tests and in some of the examples.
Converted them to fix_local and unfix_local.

> * The `docs/source/optimizable.rst` docs page still talks about `fix_all()` and `unfix_all()` rather than the new local/full functions.
Fixed this.

> * I see a few remaining `graph_` prefixes in `docs/source/mpi.rst`, `src/_core/optimizable.py`, and `examples/README.md`
Removed references to `graph_` prefixes

> * In tests/core there is both test_optimizable and test_graph_optimizable.
> * In tests/objectives there is both test_least_squares and test_graph_least_squares
> * In tests/solve there is both test_mpi and test_graph_mpi, and both test_least_squares and test_graph_least_squares
This is by design. The `test_graph_*.py` still import modules with `graph_` prefix. I want to make sure the existing code does not break down.

> * There are several orthogonal changes going on in this PR. In the future it would be nice to have these kinds of changes as separate PRs. Then we could e.g. resolve #215 without having to simultaneously merge the incomplete serialization bits.
Yes. I agree. I made the changes in separate branches and then merged them due to the anticipated merge conflicts. It is a pain point when files are renamed.  Alternatively, I should have sat on 4.

mbkumar commented:
Working implementation for serializing Optimizable objects with dependencies on other Optimizable objects.
